### PR TITLE
Parcel v2:  PostHTML transformer

### LINKED ISSUE #2159: Parcel 2: PostHTML Transform
Create the `@parcel/transform-posthtml` package in `packages/transforms/posthtml`. This package is a Parcel 2 transform plugin for PostHTML. It will be based on the `posthtml` transform in `HTMLAsset` from Parcel 1, split into its own package.

The `@parcel/transform-posthtml` package is responsible for the following things:

* Loading `.posthtmlrc` files
* Parsing and transforming if needed

### PR DESCRIPTION
**Pull request by kwelch**
# ↪️ Pull Request
Migration of the PostHTML transformer into v2.

Closes #2159.

## 🚨 Test instructions
yarn test

## ✔️ PR Todo

- [x] Added/updated unit tests for this change
- [x] Filled out test instructions (In case there aren't any unit tests)
- [x] Included links to related issues/PRs

### CODE REVIEW

**Comment by wbinnssmith** [comment]
This is so awesome!

File: packages/transformers/posthtml/.eslintrc.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (wbinnssmith) [inline packages/transformers/posthtml/.eslintrc.json]: I don't think we need this file. Was this just copied from another transformer? We should probably remove it there too at some point if so.

File: packages/transformers/posthtml/src/PostHTMLTransformer.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    const depConfig = {
+      addDependencyTo: {
+        addDependency: name =>
+          asset.addDependency(name, {includedInParent: true})
+      }
+    };
+    Object.keys(plugins).forEach(p => Object.assign(plugins[p], depConfig));
+  }
+  // TODO: find a way to load plugins
+  // config.plugins = await loadPlugins(plugins, asset.name);
```
Comment (wbinnssmith) [inline packages/transformers/posthtml/src/PostHTMLTransformer.js]: hmm not sure if we should try to make https://github.com/parcel-bundler/parcel/blob/425885992156b6b3e3ca02b3ffd1693fff6cbba5/packages/transformers/postcss/src/loadPlugins.js#L6 more generic or not

File: packages/transformers/posthtml/src/PostHTMLTransformer.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+import semver from 'semver';
+
+const getPostHTMLConfig = async asset => {
+  let config = await asset.getConfig(
+    ['.posthtmlrc', '.posthtmlrc.js', 'posthtml.config.js'],
+    {
+      packageKey: 'posthtml'
+    }
+  );
+  if (!config && !asset.options.minify) {
```
Comment (wbinnssmith) [inline packages/transformers/posthtml/src/PostHTMLTransformer.js]: We should move minifying into an optimizer (this can definitely be another PR), so can just remove these lines here.

An aside: is `options` a thing on asset? I think we should use the parcel `options` passed into all the transformer methods. But again minify stuff should be in an optimizer.

**Comment by wbinnssmith** [comment]
Thanks so much for this 😃I just merged #3062 a few hours ago which will probably create conflicts for this. Resolving should be pretty straightforward but I'd be happy to take care of that if you'd like.

File: packages/transformers/posthtml/.eslintrc.json (hunk 1 above)
Comment (kwelch) [reply packages/transformers/posthtml/.eslintrc.json]: Yes, it was before I knew what I was doing 😂. I will remove it.

File: packages/transformers/posthtml/src/PostHTMLTransformer.js (hunk 2 above)
Comment (kwelch) [reply packages/transformers/posthtml/src/PostHTMLTransformer.js]: They were shared before but I was hesitate and didn't see postcss was converted yet, so I decided to hold off on it. However I also realized this is probably why the tests are failing. 😅

File: packages/transformers/posthtml/src/PostHTMLTransformer.js (hunk 3 above)
Comment (kwelch) [reply packages/transformers/posthtml/src/PostHTMLTransformer.js]: I will remove, good call. This was directly ported and I must have glazed over this part.

File: packages/transformers/html/.eslintrc.json [hunk 4]
```diff
@@ -1,6 +0,0 @@
-{
```
Comment (wbinnssmith) [inline packages/transformers/html/.eslintrc.json]: thanks 😄

File: packages/transformers/posthtml/src/PostHTMLTransformer.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    return;
+  }
+
+  config = config || {};
+  const plugins = config.plugins;
+  if (typeof plugins === 'object') {
+    // This is deprecated in favor of result messages but kept for compatibility
```
Comment (wbinnssmith) [inline packages/transformers/posthtml/src/PostHTMLTransformer.js]: I could be wrong, but it looks like we won't need this anymore as `messages` covers the same thing, see #2966 and its changed files: https://github.com/parcel-bundler/parcel/pull/2966/files#diff-2dcba7b85fdbaf96d8c8a1e09d2c940eL39

cc @DeMoorJasper

Comment (DeMoorJasper) [reply packages/transformers/posthtml/src/PostHTMLTransformer.js]: Yeah would be great to get this deprecated feature out in Parcel 2

Comment (kwelch) [reply packages/transformers/posthtml/src/PostHTMLTransformer.js]: I will work to find a way to remove this. Wasn't care enough in to understand it when I grabbed that learning more and more about this now.

Comment (kwelch) [reply packages/transformers/posthtml/src/PostHTMLTransformer.js]: Looks like the solution, is to just remove it 😅simple enough.

**Comment by kwelch** [comment]
Been playing with this for a while I think it is close, but for some odd reason the plugin include is not finding a match in the tree. Dinner time so I am going to do the whole walk away and ✨everything works when I look again. 🤞

File: packages/transformers/posthtml/src/PostHTMLTransformer.js [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+  async getConfig({asset}) {
+    return getPostHTMLConfig(asset);
+  },
+
+  canReuseAST({ast}) {
+    return ast.type === 'posthtml' && semver.satisfies(ast.version, '^0.4.0');
+  },
+
+  async parse({asset, config}) {
+    return {
```
Comment (devongovett) [inline packages/transformers/posthtml/src/PostHTMLTransformer.js]: we should skip parsing altogether if there is no config.

File: packages/transformers/posthtml/src/PostHTMLTransformer.js [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+  async transform({asset, config}) {
+    if (!config) {
+      return [asset];
+    }
+
+    let res = await posthtml(config.plugins).process(asset.ast, config);
+
+    asset.ast = res.tree;
+    asset.isAstDirty = true;
```
Comment (devongovett) [inline packages/transformers/posthtml/src/PostHTMLTransformer.js]: I don't think you need this. Was an optimization when PostHTML was part of the HTML transform, but as a separate transformer it will already call generate when needed.

File: packages/configs/default/index.json [hunk 8]
```diff
@@ -7,7 +7,10 @@
       "@parcel/transformer-terser"
     ],
     "*.css": ["@parcel/transformer-css"],
-    "*.{htm,html}": ["@parcel/transformer-html"],
+    "*.{htm,html}": [
+      "@parcel/transformer-posthtml",
```
Comment (kwelch) [inline packages/configs/default/index.json]: I think the more aggressive thing should be first, let me know if that is not right

File: packages/transformers/html/src/HTMLTransformer.js [hunk 9]
```diff
@@ -11,13 +11,13 @@ import extractInlineAssets from './inline';
 
 export default new Transformer({
   canReuseAST({ast}) {
-    return ast.type === 'posthtml' && semver.satisfies(ast.version, '^0.4.0');
+    return ast.type === 'posthtml' && semver.satisfies(ast.version, '^0.11.3');
```
Comment (kwelch) [inline packages/transformers/html/src/HTMLTransformer.js]: The version used before was the `parser` version, but when we parse it gives us back the version of `posthtml`, so I am updating these to this.

File: packages/transformers/posthtml/src/PostHTMLTransformer.js (hunk 6 above)
Comment (kwelch) [reply packages/transformers/posthtml/src/PostHTMLTransformer.js]: Agreed!

File: packages/core/integration-tests/test/posthtml.js [hunk 10]
```diff
@@ -55,7 +59,7 @@ describe.skip('posthtml', function() {
     assert(asset.dependencies.get(other).includedInParent);
```
Comment (kwelch) [inline packages/core/integration-tests/test/posthtml.js]: I wasn't happy to skip these, but I wasn't sure the new way to do this.

File: packages/transformers/posthtml/src/loadPlugins.js [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    return [];
+  }
+}
+
+async function loadPlugin(
+  pluginArg: string | Function,
+  relative: FilePath,
+  options: mixed = {}
+): mixed {
+  if (typeof pluginArg !== 'string') {
```
Comment (kwelch) [inline packages/transformers/posthtml/src/loadPlugins.js]: I took this whole file from #3037, just had to add this line in case the config had a require statement. As @wbinnssmith mentioned, we could find a common place for this to live. Until then I duplicated it.

File: flow-libs/posthtml.js.flow [hunk 12]
```diff
@@ -14,17 +14,28 @@ declare module 'posthtml' {
     content?: Array<string>
   };
 
+  declare type PostHTMLTree = Array<PostHTMLNode>;
```
Comment (wbinnssmith) [inline flow-libs/posthtml.js.flow]: 🤩

File: packages/configs/default/index.json (hunk 8 above)
Comment (wbinnssmith) [reply packages/configs/default/index.json]: Yeah, I think I got this backwards in the PostCSS transform

File: packages/transformers/html/src/HTMLTransformer.js (hunk 9 above)
Comment (wbinnssmith) [reply packages/transformers/html/src/HTMLTransformer.js]: does the parser result actually include version information? Last I remember the posthtml ast is basically unversioned (I don't think they've made any breaking changes so far either).

Personally it doesn't matter to me as long as we commit to something and make sure third party plugins follow it.

Comment (kwelch) [reply packages/transformers/html/src/HTMLTransformer.js]: I looked into it and I didn't see a version returned from `posthtml` or the `parser` so I reverted back to using the parser version, as it felt more right.

File: packages/configs/default/index.json (hunk 8 above)
Comment (devongovett) [reply packages/configs/default/index.json]: Yeah posthtml/postcss should be first, in case they introduce new dependencies.

File: packages/configs/default/package.json [hunk 13]
```diff
@@ -22,6 +22,7 @@
     "@parcel/transformer-babel": "^2.0.0-alpha.0",
     "@parcel/transformer-css": "^2.0.0-alpha.0",
     "@parcel/transformer-html": "^2.0.0-alpha.0",
+    "@parcel/transformer-posthtml": "^2.0.0",
```
Comment (devongovett) [inline packages/configs/default/package.json]: Can you make this use an alpha version?

File: packages/transformers/posthtml/src/PostHTMLTransformer.js [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+import {Transformer} from '@parcel/plugin';
+
+import posthtml from 'posthtml';
+import parse from 'posthtml-parser';
+import render from 'posthtml-render';
+import nullthrows from 'nullthrows';
+import semver from 'semver';
+import loadPlugins from './loadPlugins';
+
+const getPostHTMLConfig = async asset => {
```
Comment (devongovett) [inline packages/transformers/posthtml/src/PostHTMLTransformer.js]: any reason why this function isn't inlined below?

File: packages/configs/default/package.json (hunk 13 above)
Comment (kwelch) [reply packages/configs/default/package.json]: Is there a "fancy" way of doing that, or should I just update version, manually?

File: packages/configs/default/index.json (hunk 8 above)
Comment (kwelch) [reply packages/configs/default/index.json]: Should I update postcss in this or separate?

File: packages/core/integration-tests/test/posthtml.js (hunk 10 above)
Comment (devongovett) [reply packages/core/integration-tests/test/posthtml.js]: Yeah I think you'd have to look for a `connected_file` node in the asset graph, but that is not exposed by Parcel. cc. @wbinnssmith any ideas?

File: packages/configs/default/package.json (hunk 13 above)
Comment (devongovett) [reply packages/configs/default/package.json]: I think just manually here and in the posthtml package.json

File: packages/transformers/posthtml/src/PostHTMLTransformer.js (hunk 14 above)
Comment (kwelch) [reply packages/transformers/posthtml/src/PostHTMLTransformer.js]: I had considered putting it in a `config.js`. I agree input would be better.

Comment (kwelch) [reply packages/transformers/posthtml/src/PostHTMLTransformer.js]: Consolidated this.

File: packages/configs/default/index.json (hunk 8 above)
Comment (kwelch) [reply packages/configs/default/index.json]: I moved `postcss` to be consistent.

Comment (kwelch) [reply packages/configs/default/index.json]: I ended up reverting the move of postcss, because it cause test failures. If it is suppose to be postcss first, we may have to look into why those tests were failing.

File: packages/core/integration-tests/test/posthtml.js (hunk 10 above)
Comment (wbinnssmith) [reply packages/core/integration-tests/test/posthtml.js]: There hasn't been a real usecase for it yet, but I'm okay exposing `getConnectedFiles()` on the public Asset object.

Comment (kwelch) [reply packages/core/integration-tests/test/posthtml.js]: Nice! I will look into exposing that in a different PR. It will be good exposure into areas outside of the transformer.

Comment (kwelch) [reply packages/core/integration-tests/test/posthtml.js]: I did update the posthtml transformer to include dependencies so that `getConnectedFile` can be eventually used.

File: packages/transformers/posthtml/src/PostHTMLTransformer.js [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+    if (!config) {
+      return [asset];
+    }
+
+    const ast = nullthrows(asset.ast);
+
+    let res = await posthtml(config.plugins).process(ast.program, config);
+
+    if (res.messages) {
+      await Promise.all(
```
Comment (devongovett) [inline packages/transformers/posthtml/src/PostHTMLTransformer.js]: doesn't look like this needs to be async right?

Comment (kwelch) [reply packages/transformers/posthtml/src/PostHTMLTransformer.js]: `asset.addConnectedFile` returns a promise, so I thought they needed to be awaited.

Comment (devongovett) [reply packages/transformers/posthtml/src/PostHTMLTransformer.js]: oh you're right. I forgot about that.
